### ISSUE TITLE
Consuming Services - How to use Service Binding based Destination access with multitenancy?

### ISSUE THREAD

**Issue by KarthikeyanLoganathan** [SAP/cloud-sdk-js#5157]
Refer to SAP internal CAP issues/issues/17158

I managed to resolve service binding destination, handling xsuaa uaasubdomain of the consumer tenant, and client credentials flow for the consumer tenant

```js
    async resolveDestinationByServiceInstanceName(cds, serviceInstanceName, tenantContext) {
        //TODO: This only supports client credentials flow
        if (cds.env?.production) {
            // need request context here.  we need tenant-id
            // console.log('cds.User: ', JSON.stringify(that.cds.User)); //id, roles, attr, tokenInfo
            // console.log('req: ', JSON.stringify(req));
            // console.log('req.authInfo: ', JSON.stringify(req.authInfo));
            //TODO: decide about caching?
            //TODO: What about time-bound caching?
            //TODO: What about user specific tokens?
            // https://sap.github.io/cloud-sdk/docs/js/features/connectivity/destinations#service-binding-environment-variables
            const destinationData = await destinationForServiceBinding(serviceInstanceName,
                {
                    /**
                     * @param {import('@sap-cloud-sdk/core').ServiceBinding} serviceBinding 
                     * @returns {import('@sap-cloud-sdk/core').Destination} Destination
                     */
                    serviceBindingTransformFn: async (serviceBinding) => {
                        //TODO: customer UAA subaccount sub-domain to be used as prefix to UAA URL, for 
                        //customer tenant specific JWT token
                        //Probably we are resolving the destination too early here
                        //At this point when server is loaded/up, you dont have end user context
                        //Only in customer tenant context, we should resolve the destination
                        //Then only it will work for all customer tenants.
                        console.log("serviceBindingTransformFn called now");
                        let uaaSubDomain = serviceBinding.credentials.uaa.identityzone;
                        const uaaCredentials = {};
                        Object.assign(uaaCredentials, serviceBinding.uaa);
                        if (tenantContext?.req) {
                            const inputToken = new xssec.XsuaaToken(tenantContext?.req?.headers?.authorization?.split(/^bearer /i)[1]);
                            uaaSubDomain = inputToken.payload.ext_attr.zdn;
                        } else if (tenantContext?.uaaSubDomain) {
                            uaaSubDomain = tenantContext?.uaaSubDomain;
                        }
                        uaaCredentials.url = `https://${uaaSubDomain}.${uaaCredentials.uaadomain}`;
                        const uaaService = new xssec.XsuaaService(uaaCredentials);
                        return {
                            name: serviceInstanceName,
                            authentication: 'OAuth2ClientCredentials',
                            clientId: uaaCredentials.clientid,
                            clientSecret: uaaCredentials.clientsecret,
                            url: serviceBinding.credentials.url,
                            tokenServiceUrl: uaaCredentials.url + '/oauth/token',
                            authTokens: [await uaaService.fetchClientCredentialsToken()].map(mapJWT)
                        }
                    }
                }
            )
            await registerDestination(destinationData, {
                useCache: false,
                IsolationStrategy: 'tenant'
            });
        }
    }

    async resolveDestinationByServiceName(cds, serviceLabel, tenantContext) {
        if (cds.env?.production) {
            const services = xsenv.filterServices({ label: serviceLabel });
            if (services.length === 0) {
                throw new Error(`Application has no service binding for label ${serviceLabel}`);
            }
            await this.resolveDestinationByServiceInstanceName(cds, services[0].name, tenantContext);
        }
    }
```

```js
    this.connectToOdataService = async function (tenantContext) {
        if (rdaFacade.rdaConfig.applicationType === rdaGlobals.ApplicationTypes.consumer) {
            await rdaFacade.resolveDestinationByServiceInstanceName(cds, "rda-consumer-product-rating-rda-service-instance", tenantContext);
        }
        return await cds.connect.to(serviceInfo.serviceId);
    };
```

the jwt token also has expiry information for sure.

I naturally expected destination service to obtain token, cache the token till expiry etc. (which currently I have coded for token fetch)

Can this work out of the box from cloud sdk or destination service?

Can we have such tenant specific service binding destinations coded for dynamic fetch of jwt, automatic caching?

**Comment by deekshas8** [comment]
Hi @KarthikeyanLoganathan ,

You can take a look at our documentation of service bindings [here](https://sap.github.io/cloud-sdk/docs/js/features/connectivity/destinations#service-binding-environment-variables).

**Comment by KarthikeyanLoganathan** [comment]
The following worked for me.

Service Binding information in VCAP_SERVICES

```json
{
    "the-service-name": [
        {
            "label": "the-service-name",
            "provider": null,
            "plan": "default",
            "name": "the-service-instance-name",
            "tags": [],
            "instance_guid": "7a1abc04-de92-4db1-b6e0-90909090909",
            "instance_name": "the-service-instance-name",
            "binding_guid": "b884a27d-406f-47bd-bb39-_____________",
            "binding_name": null,
            "credentials": {
                "url": "https://the-service-app-backend.cert.cfapps.DC.hana.ondemand.com/",
                "uaa": {
                    "tenantmode": "shared",
                    "sburl": "https://internal-xsuaa.authentication.eu12.hana.ondemand.com",
                    "subaccountid": "6c124242-103c-4db4-8fe4-xxxxxxxxxxxx",
                    "credential-type": "binding-secret",
                    "clientid": "sb-SECRET...",
                    "xsappname": "SECRET....",
                    "clientsecret": "SECRET",
                    "serviceInstanceId": "7a1abc04-de92-4db1-b6e0-012345678901",
                    "url": "https://this-sub-account-identity-zone.authentication.eu12.hana.ondemand.com",
                    "uaadomain": "authentication.eu12.hana.ondemand.com",
                    "verificationkey": "-----BEGIN PUBLIC KEY-----\n----SECRET------\n-----END PUBLIC KEY-----",
                    "apiurl": "https://api.authentication.DC.hana.ondemand.com",
                    "identityzone": "this-sub-account-identity-zone",
                    "identityzoneid": "6c124242-103c-4db4-8fe4-xxxxxxxxxxxx",
                    "tenantid": "6c124242-103c-4db4-8fe4-xxxxxxxxxxxx",
                    "zoneid": "6c124242-103c-4db4-8fe4-xxxxxxxxxxxx"
                }
            },
            "syslog_drain_url": null,
            "volume_mounts": []
        }
    ]
}
```



```javascript
const { getDestinationFromServiceBinding, serviceToken, decodeJwt, registerDestination } = require('@sap-cloud-sdk/connectivity');

    async registerDestinationDestinationFromServiceBinding(cds, serviceInstanceName) {
        if (cds.env?.production) {
            if (!this.#serviceBindingDestinationsRegistered?.[serviceInstanceName]) {
                const destination = await getDestinationFromServiceBinding({
                    destinationName: serviceInstanceName,
                    useCache: true,
                    serviceBindingTransformFn: async (service, options) => {
                        const myService = {};
                        Object.assign(myService, service);
                        delete myService.credentials;
                        myService.credentials = {};
                        Object.assign(myService.credentials, service.credentials.uaa);
                        myService.credentials.serviceurls = {url: service.credentials.url};
                        const token = await serviceToken(myService, options);
                        const exp = decodeJwt(token).exp;
                        return {
                            url: service.credentials.url,
                            authentication: 'OAuth2ClientCredentials',
                            authTokens: [{
                                value: token,
                                type: 'bearer',
                                expiresIn: exp && Math.floor((exp * 1000 - Date.now()) / 1000).toString(10) || '0',
                                http_header: { key: 'Authorization', value: `Bearer ${token}` },
                                error: null
                            }]
                        };
                    }
                });
                await registerDestination(destination);
                this.#serviceBindingDestinationsRegistered[serviceInstanceName] = {};
            }
        }
    }

    async ensureRegistrationOfRdaServiceDestinationForRdaConsumer(cds) {
        if (rdaConfig.applicationType === rdaGlobals.ApplicationTypes.consumer) {
            await this.registerDestinationDestinationFromServiceBinding(cds, rdaConfig.sourceServiceInstanceName);
        }
    }
```


I would prefer not to send authTokens, but register destination with client-id, client-secret, authentical url ... in line with documentation https://help.sap.com/docs/connectivity/sap-btp-connectivity-cf/oauth-client-credentials-authentication   

Here wish is that {tenant} place holder is handled automatically and we get token as need orchestrated by destination service.

**Comment by tomfrenken** [comment]
Generally speaking, instead of targeting a service binding you can always create the same destination in the destination service, and then target that destination instead without any issues.
Multi-tenancy there should also be no problem, if you follow the typical provider/subscriber model.

**Comment by KarthikeyanLoganathan** [comment]
I revised my comment above https://github.com/SAP/cloud-sdk-js/issues/5157#issuecomment-2481387286

**Comment by KarthikeyanLoganathan** [comment]
> I can see that you essentially rebuild [our internal function](https://github.com/SAP/cloud-sdk-js/blob/main/packages/connectivity/src/scp-cf/destination/service-binding-to-destination.ts#L95) for building a client credentials destination.

Yes, i happened to build something like [cloud-sdk internal function](https://github.com/SAP/cloud-sdk-js/blob/main/packages/connectivity/src/scp-cf/destination/service-binding-o-destination.ts#L95)

> I suppose if we offered a generic transformer function like `serviceBindingToClientCredentialsDestination` it would solve your issue?

It can help for sure.

However, my requirement is bit more than this.

* Can I register the service binding destination only once after starting my application ?
  * at that point in time, I don't want to resolve the token
  * I want to handover the following information as per documentation https://help.sap.com/docs/connectivity/sap-btp-connectivity-cf/oauth-client-credentials-authentication
    * Name = name-of-my-service-instance
    * Type = HTTP
    * URL = target-application-url-from-service-binding
    * ProxyType = Internet 
    * Authentication = OAuth2ClientCredentials 
    * clientId = client-id-from-uaa-section-of-service-binding
    * clientSecret = client-secret-from-uaa-section-of-service-binding
    * tokenServiceURL = https://{tenant}.authentication.LANDSCAPE.hana.ondemand.com/oauth/token
      * Here the {tenant} to be late resolved based on CAP runtime context, current tenant-id, current-uaa-sub-domain value to be used as {tenant}
    * tokenServiceURLType = Common 
    * ...


At the time of actual destination usage, we wish the tenant context to be propagated to this destination above
* in my case connect to remote odata service by CAP using 
  * const remoteService = cds.connect.to('odata-service-id', options)
* cds.context to have tenant, user info (user info irrelevant for client credentials flow)
* req.headers[authorization] (bearer token) to given tenant context (in case of the current inbound call is not protected/restricted)
* cds runtime picks the tenant information
* consumer application invokes 
  * remoteService.run(cql)
* Here expectation is that the tenant context/uaa-sub-domain gets propagated to cloud-sdk-js
* At this time, we expect the tenant specific token to be fetched/cached in cloud-sdk-js

**Comment by KarthikeyanLoganathan** [comment]
I see that registerDestination(destination) accepts all the information as per https://help.sap.com/docs/connectivity/sap-btp-connectivity-cf/oauth-client-credentials-authentication
* name = name-of-my-service-instance
* url = target-application-url-from-service-binding
* type = "HTTP"
* authentication = OAuth2ClientCredentials
* proxyType = Internet
* clientId = client-id-from-uaa-section-of-service-binding
* clientSecret = client-secret-from-uaa-section-of-service-binding
* tokenServiceURL = https://{tenant}.authentication.LANDSCAPE.hana.ondemand.com/oauth/token
   * **Here can the {tenant} get late resolved based on CAP runtime context, current tenant-id, current-uaa-sub-domain value to be used as {tenant}?**


only one destination to be registered in consumer application lifecycle; based on the HTTP request scope/cap context/current-runtime-tenant, {tenant} to be resolved to uaa-sub-domain;   And token to be fetched dynamically, token to be cached specific to consumer tenant.

**Comment by tomfrenken** [comment]
Hi @KarthikeyanLoganathan,

the `registerDestination` function is actually not supposed to be used in your case.
It is meant for local testing or environments where authentication is taken care of by another entity. 
For example, in Kyma, you have a sidecar, or for on-premise, there is the transparent proxy instead.

For your use case, the `getDestinationFromServiceBinding` should be all you need to use, the destinations are already cached based on the auth token duration, so you don't have to handle it.

Regarding the token service URL, e.g. `https://{tenant}.authentication.LANDSCAPE.hana.ondemand.com/oauth/token`, this is used by the app router to exchange tokens, you shouldn't have to do anything with it.

So all in all, a generic transformer function would most likely be all you need to simplify the flow that you've already implemented.

**Comment by KarthikeyanLoganathan** [comment]
input by @tomfrenken 
https://github.com/SAP/cloud-sdk-js/pull/5195/files

**Comment by KarthikeyanLoganathan** [comment]
**input by @tomfrenken**


The tokenServiceUrl is actually not needed for the SDK, it is a property used by the app router when exchanging tokens between your OAuth provider (e.g. XSUAA) and the destination service.
The resulting User JWT is used as part of the serviceToken call to get a client credentials token that is scoped for a specific user (for multi-tenant scenarios) 

You do need to make the serviceToken call still, I've created a PR that shows an implimentation for this, will be available in the next release (within this week)
 
 https://github.com/SAP/cloud-sdk-js/pull/5195/files

**Comment by KarthikeyanLoganathan** [comment]
Updated my implementation as follows.

```javascript
    async registerDestinationDestinationFromServiceBinding(cds, serviceInstanceName, req) {
        if (cds.env?.production) {
            const tenantDestinationKey = `${req.context.tenant}::${serviceInstanceName}`;
            if (serviceBindingDestinationsRegistered.hasOwnProperty(tenantDestinationKey)) {
                const expirationTime = serviceBindingDestinationsRegistered[tenantDestinationKey];
                if (((new Date()).getTime() - 100000) > expirationTime) { //if token will expire in next 100 seconds, obtain new token
                    delete serviceBindingDestinationsRegistered[tenantDestinationKey];
                }
            }
            if (!serviceBindingDestinationsRegistered?.[tenantDestinationKey]) {
                let decodedToken = null;
                const destination = await getDestinationFromServiceBinding({
                    destinationName: serviceInstanceName,
                    useCache: true,
                    jwt: req.context.user.tokenInfo.jwt,
                    isolationStrategy: 'tenant',
                    serviceBindingTransformFn: async (service, options) => {
                        // const myService = {
                        //     name: service.name,
                        //     label: service.label,
                        //     tages: [...service.tags],
                        //     credentials: {
                        //         serviceurls: {
                        //             url: service.credentials.url
                        //         },
                        //         tenantmode: service.credentials.uaa.tenantmode,
                        //         sburl: service.credentials.uaa.sburl,
                        //         subaccountid: req.context.tenant,
                        //         "credential-type": service.credentials.uaa["credential-type"],
                        //         clientid: service.credentials.uaa.clientid,
                        //         xsappname: service.credentials.uaa.xsappname,
                        //         clientsecret: service.credentials.uaa.clientsecret,
                        //         serviceInstanceId: service.credentials.uaa.serviceInstanceId,
                        //         url: `https://${req.context.user.tokenInfo.payload.ext_attr.zdn}.${service.credentials.uaa.uaadomain}/oauth/token`,
                        //         uaadomain: service.credentials.uaa.uaadomain,
                        //         verificationkey: service.credentials.uaa.verificationkey,
                        //         apiurl: service.credentials.uaa.apiurl,
                        //         identityzone: req.context.user.tokenInfo.payload.ext_attr.zdn,
                        //         identityzoneid: req.context.tenant,
                        //         tenantid: req.context.tenant,
                        //         zoneid: req.context.tenant,
                        //     }
                        // };
                        const myService = {};
                        Object.assign(myService, service);
                        delete myService.credentials;
                        myService.credentials = {};
                        Object.assign(myService.credentials, service.credentials.uaa);
                        myService.credentials.serviceurls = { url: service.credentials.url };
                        myService.credentials.url = `https://${req.context.user.tokenInfo.payload.ext_attr.zdn}.${service.credentials.uaa.uaadomain}/oauth/token`;
                        // Confirmed through ticket https://github/issues/10, https://github.com/SAP/cloud-sdk-js/issues/5157
                        // @tomfrenken The resulting User JWT is used as part of the serviceToken call to get a client credentials token 
                        // that is scoped for a specific user (for multi-tenant scenarios).  You do need to make the serviceToken call.
                        // registerDestination( ) call is not required.   getDestinationFromServiceBinding( ) already registers destination
                        // CallStack confimed that authTokens is must in the resolved destination.
                        //   getAuthenticationRelatedHeaders (node_modules/@sap-cloud-sdk/connectivity/dist/scp-cf/authorization-header.js:82)
                        //   buildAuthorizationHeaders (node_modules/@sap-cloud-sdk/connectivity/dist/scp-cf/authorization-header.js:131)
                        //   buildHeadersForDestination (node_modules/@sap-cloud-sdk/connectivity/dist/scp-cf/header-builder-for-destination.js:12)
                        //   buildHeaders (node_modules/@sap-cloud-sdk/http-client/dist/http-client.js:294)
                        //   buildHttpRequest (node_modules/@sap-cloud-sdk/http-client/dist/http-client.js:63)
                        const token = await serviceToken(myService, options);
                        decodedToken = decodeJwt(token);
                        return {
                            name: serviceInstanceName,
                            url: service.credentials.url,
                            type: 'HTTP',
                            authentication: 'OAuth2ClientCredentials',
                            proxyType: 'Internet',
                            clientId: service.credentials.uaa.clientid,
                            clientSecret: service.credentials.uaa.clientsecret,
                            authTokens: [{
                                value: token,
                                type: 'bearer',
                                expiresIn: decodedToken.exp && Math.floor((decodedToken.exp * 1000 - Date.now()) / 1000).toString(10) || '0',
                                http_header: { key: 'Authorization', value: `Bearer ${token}` },
                                error: null
                            }]
                        };
                    }
                });
                serviceBindingDestinationsRegistered[tenantDestinationKey] = decodedToken.exp * 1000;
            }
        }
    }

    async ensureRegistrationOfRdaServiceDestinationForRdaConsumer(cds, req) {
        if (rdaConfig.applicationType === rdaGlobals.ApplicationTypes.consumer) {
            await this.registerDestinationDestinationFromServiceBinding(cds, rdaConfig.sourceServiceInstanceName, req);
        }
    }

```

**Comment by KarthikeyanLoganathan** [comment]
Thanks @tomfrenken 

Till the fix is available, we have a working solution - as per your last recommendations.

```javascript
    async registerDestinationDestinationFromServiceBinding(cds, serviceInstanceName, req) {
        if (cds.env?.production) {
            const tenantDestinationKey = `${req.context.tenant}::${serviceInstanceName}`;
            if (serviceBindingDestinationsRegistered.hasOwnProperty(tenantDestinationKey)) {
                const expirationTime = serviceBindingDestinationsRegistered[tenantDestinationKey];
                if (((new Date()).getTime() - 100000) > expirationTime) { //if token will expire in next 100 seconds, obtain new token
                    delete serviceBindingDestinationsRegistered[tenantDestinationKey];
                }
            }
            if (!serviceBindingDestinationsRegistered?.[tenantDestinationKey]) {
                let decodedToken = null;
                const destination = await getDestinationFromServiceBinding({
                    destinationName: serviceInstanceName,
                    useCache: true,
                    jwt: req.context.user.tokenInfo.jwt,
                    isolationStrategy: 'tenant',
                    serviceBindingTransformFn: async (service, options) => {
                        // const myService = {
                        //     name: service.name,
                        //     label: service.label,
                        //     tages: [...service.tags],
                        //     credentials: {
                        //         serviceurls: {
                        //             url: service.credentials.url
                        //         },
                        //         tenantmode: service.credentials.uaa.tenantmode,
                        //         sburl: service.credentials.uaa.sburl,
                        //         subaccountid: req.context.tenant,
                        //         "credential-type": service.credentials.uaa["credential-type"],
                        //         clientid: service.credentials.uaa.clientid,
                        //         xsappname: service.credentials.uaa.xsappname,
                        //         clientsecret: service.credentials.uaa.clientsecret,
                        //         serviceInstanceId: service.credentials.uaa.serviceInstanceId,
                        //         url: `https://${req.context.user.tokenInfo.payload.ext_attr.zdn}.${service.credentials.uaa.uaadomain}/oauth/token`,
                        //         uaadomain: service.credentials.uaa.uaadomain,
                        //         verificationkey: service.credentials.uaa.verificationkey,
                        //         apiurl: service.credentials.uaa.apiurl,
                        //         identityzone: req.context.user.tokenInfo.payload.ext_attr.zdn,
                        //         identityzoneid: req.context.tenant,
                        //         tenantid: req.context.tenant,
                        //         zoneid: req.context.tenant,
                        //     }
                        // };
                        const myService = {};
                        Object.assign(myService, service);
                        delete myService.credentials;
                        myService.credentials = {};
                        Object.assign(myService.credentials, service.credentials.uaa);
                        myService.credentials.serviceurls = { url: service.credentials.url };
                        // myService.credentials.url = `https://${req.context.user.tokenInfo.payload.ext_attr.zdn}.${service.credentials.uaa.uaadomain}`; 
                        //no need to manipulate the credentials.url; 

                        //zid = tenant-id gets passed as http-header 'x-zid' during fetch of client credentials token in serviceToken function below.
                        //@sap/xssec/srv/service/XsuaaService.js/buildTokenRequest() function passes ZID_HEADER(val x-zid) as zone-id/tenant-id value from options.zid
                        //this options.zid comes from @sap-cloud-sdk/connectivity/dist/sap-cf/xsuaa-service.js getClientCredentialsToken(service, jwt (user-jwt)) 
                        //--> zoneId: jwt_1.getTenantId(decodedJwt); xsuaaService.fetchClientCredentialsToken({tenant: undefind, zid: arg.ZoneId});

                        // Confirmed through ticket https://github.tools.sap/I014309/2024-rda/issues/10, https://github.com/SAP/cloud-sdk-js/issues/5157
                        // @tomfrenken The resulting User JWT is used as part of the serviceToken call to get a client credentials token 
                        // that is scoped for a specific user (for multi-tenant scenarios).  You do need to make the serviceToken call.
                        // registerDestination( ) call is not required.   getDestinationFromServiceBinding( ) already registers destination
                        // CallStack confimed that authTokens is must in the resolved destination.
                        //   getAuthenticationRelatedHeaders (node_modules/@sap-cloud-sdk/connectivity/dist/scp-cf/authorization-header.js:82)
                        //   buildAuthorizationHeaders (node_modules/@sap-cloud-sdk/connectivity/dist/scp-cf/authorization-header.js:131)
                        //   buildHeadersForDestination (node_modules/@sap-cloud-sdk/connectivity/dist/scp-cf/header-builder-for-destination.js:12)
                        //   buildHeaders (node_modules/@sap-cloud-sdk/http-client/dist/http-client.js:294)
                        //   buildHttpRequest (node_modules/@sap-cloud-sdk/http-client/dist/http-client.js:63)
                        const token = await serviceToken(myService, options);
                        decodedToken = decodeJwt(token);
                        return {
                            name: serviceInstanceName,
                            url: service.credentials.url,
                            type: 'HTTP',
                            authentication: 'OAuth2ClientCredentials',
                            proxyType: 'Internet',
                            clientId: service.credentials.uaa.clientid,
                            clientSecret: service.credentials.uaa.clientsecret,
                            authTokens: [{
                                value: token,
                                type: 'bearer',
                                expiresIn: decodedToken.exp && Math.floor((decodedToken.exp * 1000 - Date.now()) / 1000).toString(10) || '0',
                                http_header: { key: 'Authorization', value: `Bearer ${token}` },
                                error: null
                            }]
                        };
                    }
                });
                serviceBindingDestinationsRegistered[tenantDestinationKey] = decodedToken.exp * 1000;
            }
        }
    }

    async ensureRegistrationOfRdaServiceDestinationForRdaConsumer(cds, req) {
        if (rdaConfig.applicationType === rdaGlobals.ApplicationTypes.consumer) {
            await this.registerDestinationDestinationFromServiceBinding(cds, rdaConfig.sourceServiceInstanceName, req);
        }
    }
```

**Comment by KarthikeyanLoganathan** [comment]
fix

https://github.tools.sap/I014309/2024-rda/blob/a0234730bcf4240508812a1ce7e98cb96ae878a4/rda/util/rdaFacade.js#L428-L524
